clients/drcachesim/reader/v2p_reader.cpp

```diff
@@ -0,0 +1,116 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "v2p_reader.h"
+
+#include <cstdint>
+#include <stdint.h>
+
+#include <cstdlib>
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <unordered_map>
+
+#include "utils.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+v2p_reader_t::v2p_reader_t()
+{
+    /* Empty. */
+}
+
+bool
+v2p_reader_t::gen_v2p_map(std::string path_to_file,
+                          std::unordered_map<uint64_t, uint64_t> &v2p_map)
+{
+    std::ifstream file(path_to_file);
+    if (!file.is_open()) {
+        ERRMSG("ERROR: Failed to open %s\n", path_to_file.c_str());
+        return false;
+    }
+
+    constexpr char virtual_address_key[] = "virtual_address";
+    constexpr char physical_address_key[] = "physical_address";
+    // Assumes virtual_address 0 is reserved and not used in the application.
+    uint64_t virtual_address = 0;
+    std::string line;
+    while (std::getline(file, line)) {
+        std::size_t found = line.rfind(virtual_address_key);
+        if (found != std::string::npos) {
+            std::stringstream ss(line);
+            std::string elem;
+            if (!std::getline(ss, elem, ':')) {
+                ERRMSG("ERROR: virtual_address key not found. Found %s instead.\n",
+                       elem.c_str());
+                return false;
+            }
+            if (!std::getline(ss, elem, ':')) {
+                ERRMSG("ERROR: virtual_address value not found. Found %s instead.\n",
+                       elem.c_str());
+                return false;
+            }
+            virtual_address = std::stoull(elem, nullptr, 0);
+            continue;
+        }
+        found = line.rfind(physical_address_key);
+        if (found != std::string::npos) {
+            if (virtual_address == 0) {
+                ERRMSG("ERROR: no corresponding virtual_address for this "
+                       "physical_address.\n");
+                return false;
+            }
+            std::stringstream ss(line);
+            std::string elem;
+            if (!std::getline(ss, elem, ':')) {
+                ERRMSG("ERROR: physical_address key not found. Found %s instead.\n",
+                       elem.c_str());
+                return false;
+            }
+            if (!std::getline(ss, elem, ':')) {
+                ERRMSG("ERROR: physical_address value not found. Found %s instead.\n",
+                       elem.c_str());
+                return false;
+            }
+            uint64_t physical_address = std::stoull(elem, nullptr, 0);
+            v2p_map[virtual_address] = physical_address;
```

Comment:
if virtual_address is 0, return an error?

Reply:
Done above, but moved the check here.